### PR TITLE
Implement TCP server Automaton

### PR DESCRIPTION
**Pull request by jabbate19**
As a child class of TCP_client, new TCP_server automatically completes a TCP handshake when a client connects and is integrated with supersocket. Inspiration was taken from #2409 

Anything that could be removed due to parent class was removed. States were required to stay due to Automaton design.

While working on this, I found/fixed a bug within TCP_client that would cause the ACK number of a response from data to be wrong when a small amount of data was sent. Padding was being considered in the length for the ack number when it is not in the protocol.

Part of #399

### CODE REVIEW

File: scapy/layers/inet.py [hunk 1]
```diff
@@ -2053,6 +2053,150 @@ def stop_ack_timeout(self):
...
+            raise self.ESTABLISHED().action_parameters(pkt)
+
+    @ATMT.receive_condition(ESTABLISHED)
+    def incoming_data_received(self, pkt):
+        if not isinstance(pkt[TCP].payload, (NoPayload, conf.padding_layer)):
+            raise self.ESTABLISHED().action_parameters(pkt)
+
+    @ATMT.action(incoming_data_received)
+    def receive_data(self, pkt):
+        super().receive_data(pkt)
```
Comment (gpotter2) [inline scapy/layers/inet.py:2145]: This won't work on Python 2 (`super(TCP_server, self)`). Did you test it? same elsewhere
(sorry this should be the last release to support it :p)

File: scapy/layers/inet.py [hunk 2]
```diff
@@ -2053,6 +2053,150 @@ def stop_ack_timeout(self):
...
+    This automaton will handle TCP 3-way handshake.
+    """
+    def parse_args(self, ip, sport, *args, **kargs):
+        from scapy.sessions import TCPSession
+        self.dst = str(Net(ip))
+        self.sport = sport
+        self.l4 = IP(src=ip) / TCP(sport=self.sport, flags=0,
+                                   seq=random.randrange(0, 2**32))
+        self.src = self.l4.src
+        self.sack = self.l4[TCP].ack
```
Comment (gpotter2) [inline scapy/layers/inet.py:2068]: This is a bit cheap: you handled only a single client at the time. I must say I envisioned something a bit more useful, handling multiple streams.
Did you have a look at Transmission Control Block?

**Review by gpotter2 — COMMENTED** [review]
Thanks a lot for the PR & your interest in Scapy !
This is an interesting topic and it's really nice to see what you did.

Unfortunately though, I believe it would need a bit of refactoring: we want to support multiple clients, so you probably would need to implement TCBs of some sort. There is a bit of refactoring to do :p

**Comment by jabbate19** [comment]
I will continue this soon, been out of town for a bit so please don't close

**Comment by stryngs** [comment]
Still working on this @jabbate19 ?

**Comment by gpotter2** [comment]
Closing for inactivity.

**Pull request closed without merging**